custom/enikshay/management/commands/drug_resistance_reconciliation.py

```diff
@@ -0,0 +1,197 @@
+        return file_name
+
+    def writerow(self, row):
+        with open(self.result_file, 'a') as csvfile:
+            writer = csv.DictWriter(csvfile, fieldnames=self.get_result_file_headers())
+            writer.writerow(row)
+
+    @staticmethod
+    def public_app_case(occurrence_case_id):
+        person_case = get_person_case_from_occurrence(DOMAIN, occurrence_case_id)
+        person_case_properties = person_case.dynamic_case_properties()
+        if person_case_properties.get(ENROLLED_IN_PRIVATE) == 'true':
+            return False
+        return True
+
+    def reconcile_case(self, occurrence_case_id):
+        # get all open drug resistance cases
+        # group by drug_id
+        # if any drug_id has more than one corresponding drug resistance case, fix them
+        drug_resistance_cases = get_open_drug_resistance_cases_from_occurrence(occurrence_case_id)
+        drug_resistance_cases_by_drug_id = defaultdict(list)
+        for drug_resistance_case in drug_resistance_cases:
+            drug_id = drug_resistance_case.dynamic_case_properties().get('drug_id')
+            if drug_id not in self.drug_id_values:
+                print("New drug_id value found, %s" % drug_id)
+                self.drug_id_values.append(drug_id)
+            drug_resistance_cases_by_drug_id[drug_id].append(drug_resistance_case)
+        for drug_id in drug_resistance_cases_by_drug_id:
+            drug_resistance_cases_for_drug = drug_resistance_cases_by_drug_id[drug_id]
+            if len(drug_resistance_cases_for_drug) > 1:
+                self.reconcile_drug_resistance_cases(drug_resistance_cases_for_drug, drug_id, occurrence_case_id)
+
+    def reconcile_drug_resistance_cases(self, drug_resistance_cases, drug_id, occurrence_case_id):
+        """
+        Take cases with the same drug_id and then keep only one and close others.
+        Find the one in this order
+        - that has sensitivity resistant
+        - that has sensitivity sensitive
+        - the first ever opened case
+        :param drug_resistance_cases: under an occurrence with the same drug_id
+        """
+        # sanity check that we got more than one case so we should be considering closing cases
+        # and confirm that all have the same drug id
+        if (len(drug_resistance_cases) < 2 or
+            any(drug_resistance_case.dynamic_case_properties().get('drug_id') != drug_id for
+                drug_resistance_case in drug_resistance_cases)):
+            raise CommandError("Asked to reconcile cases when not needed for occurrence case %s"
+                               % occurrence_case_id)
+
+        drug_resistance_case_ids = set([case.get_id for case in drug_resistance_cases])
+        resistant_drug_resistance_cases = []
+        sensitive_drug_resistance_cases = []
+
+        # group cases by their sensitivity
+        # possible values are resistant/ sensitive/ unknown
+        for drug_resistance_case in drug_resistance_cases:
```

Comment:
@mkangia 

My idea for how to do this was to to do a multikey sort on the drug_resistance cases
- First sort descending by sensitivity (resistant = 3, sensitive = 2, anything else is 1_
-Then sort by opened_on (ascending)

Then keep the first result.  Not sure if that would be easier to parse / read.

Reply:
You could use `min` and pass a key function which returns a `(preference, opened_on)` tuple (where preference is resistant = 1, sensitive = 2, anything else is 3).  That should do the ordering properly.